lib/client/mfa.go

```diff
@@ -22,101 +22,129 @@ import (
+			fmt.Fprintln(os.Stderr) // Print a new line after the prompt
+			if err != nil {
+				respC <- response{kind: kind, err: err}
+				return
+			}
+			respC <- response{
+				kind: kind,
+				resp: &proto.MFAAuthenticateResponse{
+					Response: &proto.MFAAuthenticateResponse_TOTP{
+						TOTP: &proto.TOTPResponse{Code: code},
+					},
+				},
 			}
 		}()
+	} else {
+		fmt.Fprintf(os.Stderr, "Tap any %ssecurity key\n", promptDevicePrefix)
+	}
 
-		go func() {
-			totpCode, err := prompt.Input(ctx, os.Stderr, prompt.Stdin(), fmt.Sprintf("Tap any %[1]ssecurity key or enter a code from a %[1]sOTP device", promptDevicePrefix, promptDevicePrefix))
-			res := response{kind: "TOTP", err: err}
-			if err == nil {
-				res.resp = &proto.MFAAuthenticateResponse{Response: &proto.MFAAuthenticateResponse_TOTP{
-					TOTP: &proto.TOTPResponse{Code: totpCode},
-				}}
-			}
+	// TODO(codingllama): Make sure users get a reasonable error back when the
+	//  origin doesn't match the RPID. Webauthn isn't as flexible as U2F in this
+	//  regard.
+	origin := proxyAddr
+	if !strings.HasPrefix(origin, "https://") {
+		origin = "https://" + origin
+	}
 
-			select {
-			case resCh <- res:
-			case <-ctx.Done():
-			}
+	// Fire Webauthn or U2F goroutine.
+	switch {
+	case c.WebauthnChallenge != nil:
+		go func() {
+			log.Debugf("WebAuthn: prompting U2F devices with origin %q", origin)
+			resp, err := promptWebauthn(ctx, origin, wanlib.CredentialAssertionFromProto(c.WebauthnChallenge))
+			// Technically it's kind="Webauthn", but we are not changing CLI
+			// nomenclature yet.
+			respC <- response{kind: "U2F", resp: resp, err: err}
+		}()
+	case len(c.U2F) > 0:
+		go func() {
+			log.Debugf("prompting U2F devices with facet %q", origin)
+			resp, err := promptU2FChallenges(ctx, proxyAddr, c.U2F)
+			respC <- response{kind: "U2F", resp: resp, err: err}
 		}()
+	}
 
-		for i := 0; i < 2; i++ {
-			select {
-			case res := <-resCh:
-				if res.err != nil {
-					log.WithError(res.err).Debugf("%s authentication failed", res.kind)
-					continue
-				}
-
-				// Print a newline after the TOTP prompt, so that any future
-				// output doesn't print on the prompt line.
-				fmt.Fprintln(os.Stderr)
-
-				return res.resp, nil
-			case <-ctx.Done():
-				return nil, trace.Wrap(ctx.Err())
+	for i := 0; i < numGoroutines; i++ {
+		select {
+		case resp := <-respC:
+			if err := resp.err; err != nil {
+				log.WithError(err).Debugf("%s authentication failed", resp.kind)
+				continue
 			}
+
+			// Exiting cancels the context via defer, which makes the remaining
+			// goroutines stop.
+			return resp.resp, nil
+		case <-ctx.Done():
+			return nil, trace.Wrap(ctx.Err())
 		}
-		return nil, trace.Errorf("failed to authenticate using all U2F and TOTP devices, rerun the command with '-d' to see error details for each device")
-	default:
-		return nil, trace.BadParameter("bug: non-exhaustive switch in promptMFAChallenge")
 	}
+	return nil, trace.BadParameter(
```

Comment:
This doesn't seem like a BadParameter error to me.

Reply:
Yes, this looks more like a failed-precondition/illegal-state kind of error to me, but @r0mant tells me we default to BadParameter when the other, more specific trace errors don't apply (instead of doing a bare trace.Errorf).